net/libfab/libfab.c

```diff
@@ -118,60 +125,219 @@ M0_INTERNAL void m0_net_libfab_fini(void)
+					       "active endpoint = %d", rc);
+					libfab_ep_param_free(ep, tm);
+				}
+			}
+		} else
+			M0_LOG(M0_ERROR, "Received unwanted event = %d", event);
+	} else {
+		memset(&eq_err, 0, sizeof(eq_err));
+		rc = fi_eq_readerr(eq, &eq_err, 0);
+		if (rc != sizeof(eq_err)) {
+			M0_LOG(M0_ERROR, "fi_eq_readerr returns error =%s",
+			       fi_strerror((int) -(rc)));
+		} else {
+			M0_LOG(M0_ERROR, "fi_eq_readerr provider err no %d:%s",
+				eq_err.prov_errno,
+				fi_eq_strerror(eq, eq_err.prov_errno,
+					       eq_err.err_data, NULL, 0));
+		}
+	}
+	return 1;
+}
+
 /**
  * Used to poll for connection and completion events
  */
 static void libfab_poller(struct m0_fab__tm *tm)
 {
+	struct fi_cq_entry       comp;
+	int                      wait_cnt;
+	int                      poll_cnt;
+	void                    *ctx[8];
+	int                      i;
+	int                      rc = 0;
+	struct fi_cq_err_entry   cq_err;
+	uint32_t                 cnt;
+
+	tm_event_post(tm, M0_NET_TM_STARTED);
 	while (tm->ftm_shutdown == false) {
-		struct fi_cq_entry     comp;
-		struct fi_eq_cm_entry  entry;
-		uint32_t               event;
-		ssize_t                rd;
-		int                    wait_cnt;
-		int                    poll_cnt;
-		void                  *ctx[8];
-		int                    i;
-		int                    rc = 0;
 
 		memset(ctx, 0, sizeof(ctx));
 		wait_cnt = fi_wait(tm->ftm_waitset, -1);
-		
 		if (wait_cnt) {
-			poll_cnt = fi_poll(tm->ftm_pollset, ctx, 
+			poll_cnt = fi_poll(tm->ftm_pollset, ctx,
 					   ARRAY_SIZE(ctx));
 			for (i = 0; i < poll_cnt; i++) {
 				rc = fi_cq_read(ctx[i], &comp, 1);
-				if (rc > 0) {
-					/* TODO: Action after 
-					succesful completion */
-				} else if (rc < 0 && rc != -FI_EAGAIN) {
-					if (rc == -FI_EAVAIL) {
-						/* TODO: Decode error */
-					}
-					else {
-						/* TODO: Action if err code 
-						not available */
+				if (rc >= 0)
+					buf_done(comp.op_context, 0);
+				else {
+					/* In case of FI_EAGAIN wait for
+					* completion and do not cancel
+					*/
+					if (rc != -FI_EAGAIN) {
+						rc = fi_cq_readerr(ctx[i],
+								   &cq_err, 0);
+						if ( rc >= 0) {
+							M0_LOG(M0_ERROR,
+							"fi_cq_readerr err:%d",
+							cq_err.err);
+						}
+						buf_done(comp.op_context,
+							 -ECANCELED);
 					}
 				}
 			}
 
 			if (wait_cnt != poll_cnt) {
-				/* TODO: Check all event queues for FI_CONNREQ
-				events */
-				
-				rd = fi_eq_read(tm->ftm_pep->fep_ep_res.fer_eq,
-						&event, &entry, sizeof(entry), 
-						0);
-				if (rd != sizeof(entry)) {
-					/* TODO: Check for error */
-				}
-				if (event == FI_CONNREQ) {
-					/* TODO: Action on incoming connection
-					request */
-				}
+				cnt = 0;
+				/* Check for connection request events*/
+				cnt = libfab_handle_connect_request_events(tm);
+				/* Monitor connection established events*/
+				cnt += libfab_handle_connected_events(tm);
+				M0_ASSERT((wait_cnt - poll_cnt) >= cnt );
```

Comment:
the condition should be 
(wait_cnt - poll _cnt) <= cnt
as there can be some new events registered in the event queue between fi_wait/fi_poll call and this line